File: julia/clock/clock.jl
import Base: +, -, ==
import Dates: Minutes
using Printf

struct Clock
    hour
    minute
    function Clock(hour, minute)
        while minute < 0
            minute += 60
            hour -= 1
        end
        hour += minute ÷ 60
        minute = minute % 60
        while hour < 0
            hour += 24
        end
        new(hour % 24, minute)
    end
end

function ==(c1::Clock, c2::Clock)
    c1.hour == c2.hour && c1.minute == c2.minute
end

function +(c::Clock, m::Minute)
    Clock(c.hour, c.minute + m.value)
end

function -(c::Clock, m::Minute)
    Clock(c.hour, c.minute - m.value)
end

function Base.show(io::IO, clock::Clock)
    @printf(io, "\"%02d:%02d\"", clock.hour, clock.minute)
end

File: ruby/clock/clock.rb
class Clock

    attr_reader :hour, :minute

    def initialize(hour: 0, minute: 0)
        @minute = minute % 60
        @hour = (hour + (minute / 60)) % 24
    end

    def to_s
        "#{@hour.to_s.rjust(2, '0')}:#{@minute.to_s.rjust(2, '0')}"
    end

    def +(other_clock)
        Clock.new(hour: @hour + other_clock.hour, minute: @minute + other_clock.minute)
    end

    def -(other_clock)
        Clock.new(hour: @hour - other_clock.hour, minute: @minute - other_clock.minute)
    end

    def ==(o)
        o.hour == @hour && o.minute == @minute
    end
end